File: swordOffer/dfs/kthLargest.cc
#include <iostream>
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  int kthLargest(TreeNode* root, int k) {
    vector<int> inorderNodes;
    inorderT(inorderNodes, root);
    while (k > 1) {
      inorderNodes.pop_back();
      k--;
    }
    return inorderNodes.back();
  }

  void inorderT(vector<int>& list, TreeNode* node) {
    if (node == nullptr) {
      return;
    }

    inorderT(list, node->left);
    list.push_back(node->val);
    inorderT(list, node->right);
  }
};

File: swordOffer/go/dfs/kthLargest.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func kthLargest(root *TreeNode, k int) int {
	var res []int
	preorder(root, &res)
	return res[len(res)-k]
}

func preorder(root *TreeNode, res *[]int) { // in go, must use pointer in param if inner logic need modify slice cap and len
	if root == nil {
		return
	}
	preorder(root.Left, res)
	*res = append(*res, root.Val)
	preorder(root.Right, res)
}

/*
// this is kth smallest node....
func kthLargest(root *TreeNode, k int) int {
	res := make([]int, k)
	idx := 0
	preorder(root, res, &idx)
	return res[len(res)-1]
}

func preorder(root *TreeNode, res []int, idx *int) {
	if root == nil {
		return
	}

	preorder(root.Left, res, idx)
	if *idx < len(res) {
		res[*idx] = root.Val
		*idx++
	} else {
		return
	}
	preorder(root.Right, res, idx)
}
*/
